util/retry/retry.go

```diff
@@ -26,6 +27,82 @@ import (
 	"github.com/cockroachdb/cockroach/util/log"
 )
 
+type Config struct {
+	InitialInterval     time.Duration
+	MaxInterval         time.Duration
+	RandomizationFactor float64
+	Multiplier          float64
+	MaxRetries          int
+	Stopper             *util.Stopper
+}
+
+type Retry struct {
+	config         Config
+	currentAttempt int
+	isReset        bool
+}
+
+func Start(cfg Config) Retry {
+	if cfg.InitialInterval == 0 {
+		cfg.InitialInterval = 50 * time.Millisecond
+	}
+	if cfg.MaxInterval == 0 {
+		cfg.MaxInterval = 2 * time.Second
+	}
+	if cfg.RandomizationFactor == 0 {
+		cfg.RandomizationFactor = 0.15
+	}
+	if cfg.Multiplier == 0 {
+		cfg.Multiplier = 2
+	}
+
+	r := Retry{config: cfg}
+	r.Reset()
+	return r
+}
+
+func (r *Retry) Reset() {
+	r.currentAttempt = 0
+	r.isReset = true
+}
+
+func (r Retry) currentInterval() time.Duration {
+	interval := float64(r.config.InitialInterval) * math.Pow(r.config.Multiplier, float64(r.currentAttempt))
+	maxInterval := float64(r.config.MaxInterval)
+	if interval > maxInterval {
+		interval = maxInterval
+	}
+
+	var delta = r.config.RandomizationFactor * interval
+	// Get a random value from the range [interval - delta, interval + delta].
+	// The formula used below has a +1 because time.Duration is an int64, and the
+	// converstion floors the float64.
+	return time.Duration(interval - delta + rand.Float64()*(2*delta+1))
+}
+
+func (r *Retry) Next() bool {
+	if r.isReset {
+		r.isReset = false
+		return true
+	}
+
+	defer func() {
```

Comment:
right, it's just a matter of accounting. in this case, our attempt numbers are 0-indexed